File: backend/scoring.py
"""Efficiency (token %) and clarity heuristics for Human Delta."""

from __future__ import annotations

import re

_VAGUE = (
    "something",
    "stuff",
    "things",
    "somehow",
    "maybe",
    "might",
    "kind of",
    "sort of",
    "etc",
    "and so on",
)


def efficiency_percent(before: float, after: float) -> float:
    if before <= 0:
        return 0.0
    raw = ((before - after) / before) * 100
    return max(0.0, min(100.0, round(raw, 1)))


def clarity_score(
    original: str,
    optimized: str,
    mode: str,
    reverted: bool,
    *,
    meaning_loss: bool,
    constraint_drop: bool,
) -> float:
    """
    Net clarity / quality score 0–100 (hackathon heuristic).
    """
    score = 55.0

    lo = original.lower()
    opt = optimized.lower()

    removed = 0
    for w in _VAGUE:
        if w in lo and w not in opt:
            removed += 1
    score += min(10.0, removed * 3.5)

    if mode == "structured" and "\n- " in optimized:
        score += 5.0
    elif "\n" in optimized and "\n" not in original and mode == "precise":
        score += 5.0

    if reverted:
        score += 3.0

    if meaning_loss:
        score -= 20.0
    if constraint_drop:
        score -= 10.0

    return max(0.0, min(100.0, round(score, 1)))


def detect_meaning_loss(original: str, optimized: str, mode: str) -> bool:
    if mode == "compact":
        return False
    if not original.strip():
        return False
    ratio = len(optimized) / max(len(original), 1)
    return ratio < 0.42

File: frontend/lib/scoring.js
/**
 * Token efficiency % and clarity heuristics (parity with backend/scoring.py).
 */

const VAGUE = [
  "something",
  "stuff",
  "things",
  "somehow",
  "maybe",
  "might",
  "kind of",
  "sort of",
  "etc",
  "and so on",
];

export function tokenReductionPct(beforeTokens, afterTokens) {
  if (beforeTokens <= 0) return 0;
  const raw = ((beforeTokens - afterTokens) / beforeTokens) * 100;
  return Math.max(0, Math.min(100, Math.round(raw * 10) / 10));
}

export function detectMeaningLoss(original, optimized, mode) {
  if (mode === "compact") return false;
  if (!original.trim()) return false;
  const ratio = optimized.length / Math.max(original.length, 1);
  return ratio < 0.42;
}

/**
 * @param {boolean} opts.meaningLoss
 * @param {boolean} opts.constraintDrop
 */
export function computeClarityScore(
  original,
  optimized,
  mode,
  reverted,
  { meaningLoss, constraintDrop },
) {
  let score = 55;
  const lo = original.toLowerCase();
  const opt = optimized.toLowerCase();

  let removed = 0;
  for (const w of VAGUE) {
    if (lo.includes(w) && !opt.includes(w)) removed += 1;
  }
  score += Math.min(10, removed * 3.5);

  if (mode === "structured" && optimized.includes("\n- ")) score += 5;
  else if (mode === "precise" && optimized.includes("\n") && !original.includes("\n"))
    score += 5;

  if (reverted) score += 3;
  if (meaningLoss) score -= 20;
  if (constraintDrop) score -= 10;

  return Math.max(0, Math.min(100, Math.round(score * 10) / 10));
}
